init labplot, cantor and supporting libraries

hqurve opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

hqurve commented on pkgs/applications/kde/cantor.nix:
```diff
@@ -0,0 +1,91 @@
+
+, libspectre
+
+, backends ? [] # array of packages to provide for backends
+  # optional
+  # analitza,
+  # julia_17-bin,
+  # libqalculate,
+  # luajit,
+  # python3,
+  # R,
+  # sage,
```
I was unsure if there was a better way to provide the backends since I think it links them in the output hence causing total output to grow

SuperSandro2000 replied:
Do we link against them or just need to add them to PATH?

hqurve replied:
I do not know how to make it so that they are not linked. I inspected the strings in each of the backend binaries to see if they are linked

- `strings ./result/bin/.cantor_juliaserver-wrapped | rg /nix` contains two references to /nix/store/****-julia-1.8.4/lib
- `strings ./result/bin/.cantor_pythonserver-wrapped | rg /nix` contains a reference to /nix/store/****-python3-3.10.9/lib
- `strings ./result/bin/.cantor_rserver-wrapped | rg /nix | rg R-4.2.2` contains references to /nix/store/****-R-4.2.2/lib

(Note that I am looking at the `.***-wrapped` files due to the wraphook)

Also, for lua, libqalculate and octave,  there are no binaries in /bin.

hqurve replied:
Currently cantor startup screen shows the following
![Screenshot from 2023-01-08 15-14-03](https://user-images.githubusercontent.com/53281855/211214496-2f97e2a2-afd1-4e76-a131-950847a536d1.png)

If I comment all of the backends out in the `buildInputs` list, I have the following
![Screenshot from 2023-01-08 15-15-23](https://user-images.githubusercontent.com/53281855/211214547-0ee6383a-d63f-426b-9c82-424b9b1f0134.png)

wishfort36 commented:
Great work! Cantor not being in nixpkgs has irked me for a while. I once tried to make a build based on the previous PR, but it didn't work with sage, so I gave up.

Just a word of warning to those testing this: With the current build of sage in nixpkgs, Cantor (in this PR) doesn't work with sage. It will give errors like this:

```
>>> 1+1
The Sage process exited while evaluating this expression
The last output was:
```

The `cantor-execsage` script (in `share/cantor/sagebackend` in the build) runs `sage-ipython`, which isn't in `$PATH` in the nixpkgs build of sage. This seems to be the `sage` package(?) providing only a single executable to `$PATH`, called `sage`, which is a shell script that sets some env variables and calls `sage` from the `sage-with-env` package. Said package has a `sage-ipython` binary.

Cantor seems to work with sage when the `bin` folder in `sage-with-env` is added to the `$PATH`. The following made it work for me:

```
$ cd /nix/store/0dsvd919q9yx3i2rp649lf9fx17bvmwr-sage-with-env-9.6/bin/
$ PATH="$(pwd):$PATH"
$ # (run cantor)
```

ttuegel commented on pkgs/top-level/aliases.nix:
```diff
@@ -164,6 +164,7 @@ mapAliases ({
   calibre-py3 = throw "calibre-py3 has been removed from nixpkgs, as calibre's default python version is now 3. Please use calibre as replacement"; # Added 2021-01-13
   callPackage_i686 = pkgsi686Linux.callPackage;
   cantarell_fonts = throw "'cantarell_fonts' has been renamed to/replaced by 'cantarell-fonts'"; # Converted to throw 2022-02-22
+  cantor = libsForQt5.cantor; # Added 2022-06-30
```
The `aliases.nix` file is only for compatibility aliases. These are new packages, so there is no compatibility issue. These aliases should go in `all-packages.nix` so that they aren't deleted when they expire.

ttuegel commented on pkgs/applications/science/math/labplot/default.nix:
```diff
@@ -0,0 +1,108 @@
+  meta = with lib; {
+    description = "LabPlot is a FREE, open source and cross-platform Data Visualization and Analysis software accessible to everyone";
+    homepage = "https://labplot.kde.org";
+    license = []; # ?? is it multi licenced?
```
Yes, we should list all the relevant licenses here.

0pointerexception replied:
```
license = with licenses; [ asl20 bsd3 cc-by-30 cc0 gpl2Only gpl2Plus gpl3Only gpl3Plus lgpl3Plus mit];
```

should cover all licenses listed in the [LICENSES · master · Education / LabPlot · GitLab](https://invent.kde.org/education/labplot/-/tree/master/LICENSES) subdir of labplots upstream repository.

hqurve replied:
thanks

SuperSandro2000 commented on pkgs/top-level/aliases.nix:
```diff
@@ -685,6 +686,7 @@ mapAliases ({
 
   ### L ###
 
+  labplot = libsForQt5.labplot; # Added 2022-06-30
```
```suggestion
```
we do not add aliases for new packages

SuperSandro2000 commented on pkgs/development/libraries/readstat/default.nix:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  nativeBuildInputs = [
+    autoreconfHook
```
Why not compile from source?

hqurve replied:
readstat seems to have already been added to nixpkgs

SuperSandro2000 commented on pkgs/applications/science/math/labplot/default.nix:
```diff
@@ -0,0 +1,108 @@
+
+    syntax-highlighting
+    gsl
+
+
```
```suggestion

```

SuperSandro2000 commented on pkgs/applications/kde/cantor.nix:
```diff
@@ -0,0 +1,100 @@
+  ];
+
+  qtWrapperArgs = [
+    "--prefix PATH : ${placeholder "out"}/bin"
+    "--prefix PATH : ${sagelib}/bin"
```
```suggestion
    "--prefix PATH : ${placeholder "out"}/bin:${sagelib}/bin"
```

hqurve replied:
Thanks

hervyqa commented:
wow great work..
is this still a draft and not ready to merge yet?

hqurve commented:
> wow great work..
> is this still a draft and not ready to merge yet?

Thanks. Cantor builds fine but labplot does not (perhaps we could just have cantor for now?)

When I rebased the branch, labplot failed to build due to a change in (i believe) kcoreaddons. So, I updated labplot to 2.10 but it had an additional dependency dbc-parser-cpp. I've gotten the dependency to build, but I am unable to get the labplot build process to detect it

hqurve committed:
analitza: init at 23.04.0

hqurve committed:
sage: expose sage-with-env

hqurve committed:
cantor: init at 23.04.0

hqurve committed:
labplot: init at 2.10.0

hqurve commented:
I (gave up) and disabled the vector BLF build. Labplot builds now.

I also directly wrap the `cantor-execsage` script instead of wrapping cantor. I did this since labplot was also failing to load sagemath.

hqurve commented on pkgs/applications/kde/cantor.nix:
```diff
@@ -0,0 +1,104 @@
+  ];
+
+  postInstall = ''
+    wrapProgram $out/share/cantor/sagebackend/cantor-execsage \
+      --prefix PATH : ${sage.with-env}/bin
```
I am not sure if I should use wrapProgram or if I should just edit the file to have an absolute path to sage-ipython. The file's contents are

```
#!/nix/store/zlf0f88vj30sc7567b80l52d19pbdmy2-bash-5.2-p15/bin/bash

export SAGE_ROOT="$($1 --root)"

# Sage 9.2 stop source sage-env-config automatically, so do it manually
if [[ -f $SAGE_ROOT"/bin/sage-env" ]]; then
    if [[ -f $SAGE_ROOT"/bin/sage-env-config" ]]; then
        source "$SAGE_ROOT/bin/sage-env-config" >&2
    fi
    . "$SAGE_ROOT/bin/sage-env" >&2
elif [[ -f $SAGE_ROOT"/local/bin/sage-env" ]]; then
    if [[ -f $SAGE_ROOT"/local/bin/sage-env-config" ]]; then
        source "$SAGE_ROOT/local/bin/sage-env-config" >&2
    fi
    . "$SAGE_ROOT/local/bin/sage-env" >&2
elif [[ -f $SAGE_ROOT"/src/bin/sage-env" ]]; then
    if [[ -f $SAGE_ROOT"/src/bin/sage-env-config" ]]; then
        source "$SAGE_ROOT/src/bin/sage-env-config" >&2
    fi
    . "$SAGE_ROOT/src/bin/sage-env" >&2
fi

exec sage-ipython -i --simple-prompt
```

SuperSandro2000 commented on pkgs/applications/kde/cantor.nix:
```diff
@@ -0,0 +1,104 @@
+    libspectre
+  ]
+  # backends
+  ++ lib.optional (analitza != null) analitza
+  ++ lib.optional (julia != null) julia
+  ++ lib.optional (libqalculate != null) libqalculate
+  ++ lib.optional (luajit != null) luajit
+  ++ lib.optional (python3 != null) python3
+  ++ lib.optional (R != null) R
+  ++ lib.optional (sage-with-env != null) sage-with-env
```
Overwriting packages with null is rather ugly, booleans should be used instead

SuperSandro2000 commented on pkgs/applications/science/math/labplot/default.nix:
```diff
@@ -0,0 +1,112 @@
+, zlib
+, lz4
+, readstat
+# , liborigin # vendored
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/science/math/labplot/default.nix:
```diff
@@ -0,0 +1,112 @@
+    zlib
+    lz4
+    readstat
+    # liborigin
```
```suggestion
```

SuperSandro2000 reviewed (approved):
when I have the meta without typos I will merge this
